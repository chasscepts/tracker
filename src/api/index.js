import axios from 'axios';

const normalizeError = (err) => {
  if (!err) {
    return { message: 'An unknown error encountered. Please try again.' };
  }
  if (err.response) {
    return { message: err.response.data.message || JSON.stringify(err.response.data) };
  }
  if (err.request) {
    return { message: 'Server is not responding.' };
  }
  if (err.message) {
    return { message: err.message };
  }
  if (typeof err === 'string') {
    return { message: err };
  }
  return { message: 'An unknown error encountered. Please try again.' };
};

const instantiate = (token) => {
  const instance = axios.create({
    baseURL: 'http://localhost:3000/',
    responseType: 'json',
  });
  instance.defaults.headers.common.Authorization = `Bearer ${token}`;

  return instance;
};

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const get = (instance, path) => new Promise((resolve, reject) => {
  instance.get(path)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @param {{ string: any }} data sent in body of this post
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const post = (instance, path, data) => new Promise((resolve, reject) => {
  instance.post(path, data)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

const api = {
  getGroups: (token) => get(instantiate(token), '/groups'),
  getTasks: (token, start, end) => {
    let path = '/tasks';
    if (start) {
      path = `${path}?start=${start}`;
      if (end) {
        path = `${path}&end=${end}`;
      }
    }

    return get(instantiate(token), path);
  },
  updateTask: (token, title) => post(instantiate(token), '/tasks', { title }),
};

export default api;
