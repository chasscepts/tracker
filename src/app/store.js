import { configureStore } from '@reduxjs/toolkit';
import authReducer from '../reducers/authSlice';
import tasksReducer from '../reducers/tasksSlice';
import timerReducer from '../reducers/timerSlice';

export default configureStore({
  reducer: {
    auth: authReducer,
    tasks: tasksReducer,
    timer: timerReducer,
  },
});
