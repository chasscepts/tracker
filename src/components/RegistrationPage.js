import { useState } from 'react';
import mb from '../assets/css/MobileInput.module.css';

const styles = {
  container: {
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    padding: '30px',
  },
  innerWrap: {
    position: 'relative',
    padding: '15px',
    border: '1px solid #ddd',
    borderRadius: '8px',
  },
  controls: {
    marginTop: '30px',
    display: 'flex',
    justifyContent: 'flex-end',
  },
  btn: {
    fontWeight: 'bold,',
    fontSize: '1.4rem',
    lineHeight: 1,
    borderRadius: '4px',
    boxShadow: '0 3px 9px rgb(0 0 0 / 5%)',
    transition: 'all 180ms',
    display: 'block',
    width: '100%',
    marginTop: '30px',
    padding: '15px',
    textAlign: 'center',
    color: '#fff',
    backgroundColor: '#9e0606',
    borderWidth: 0,
  },
  link: {
    color: '#62b5e5',
    textDecoration: 'underline',
  },
};

export default function RegistrationPage() {
  const [email, setEmail] = useState('');
  const [password, setPassord] = useState('');

  const handleTextChange = (evt) => {
    const { name, value } = evt.target;
    if (name === 'email') {
      setEmail(value);
    } else if (name === 'password') {
      setPassord(value);
    }
  };

  const handleSubmit = () => {

  };

  return (
    <div style={styles.container}>
      <div style={styles.innerWrap}>
        <h2>Registration</h2>
        <input className={mb.text} type="text" name="email" value={email} placeholder="Enter Email" onChange={handleTextChange} />
        <input className={mb.text} type="password" name="password" value={password} placeholder="Enter Password" onChange={handleTextChange} />
        <button style={styles.btn} type="button" onClick={handleSubmit}>Register</button>
        <div style={styles.controls}>
          <span style={styles.link}>Login</span>
        </div>
      </div>
    </div>
  );
}
