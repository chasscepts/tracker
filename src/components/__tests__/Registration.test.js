import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import RegistrationPage from '../RegistrationPage';
import store from '../../app/store';
import { registerUser } from '../../api';

jest.mock('../../api');

function Wrapper() {
  return (
    <Provider store={store}>
      <MemoryRouter>
        <RegistrationPage />
      </MemoryRouter>
    </Provider>
  );
}

describe('Registration', () => {
  it('matches snapshot', () => {
    const tree = renderer
      .create(<Wrapper />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays the heading', () => {
    render(<Wrapper />);
    expect(screen.getByText('Registration')).toBeInTheDocument();
  });

  it('displays the login link', async () => {
    render(<Wrapper />);
    expect(screen.getByText('Login')).toBeInTheDocument();
  });

  it('calls api to register user', async () => {
    registerUser.mockResolvedValue({ user: { email: 'test@example.com', id: 1 } });

    render(<Wrapper />);
    userEvent.type(screen.getByPlaceholderText('Enter Email'), 'test.example.com');
    userEvent.type(screen.getByPlaceholderText('Enter Password'), 'password');
    await userEvent.click(screen.getByText('Register'));
    await Promise.resolve();
    expect(registerUser).toHaveBeenCalledTimes(1);
  });
});
