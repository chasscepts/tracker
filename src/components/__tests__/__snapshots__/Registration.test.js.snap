// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration matches snapshot 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "justifyContent": "center",
      "padding": "30px",
    }
  }
>
  <div
    style={
      Object {
        "border": "1px solid #ddd",
        "borderRadius": "8px",
        "padding": "15px",
        "position": "relative",
      }
    }
  >
    <h2
      style={
        Object {
          "color": "#62b5e5",
        }
      }
    >
      Registration
    </h2>
    <input
      className="text"
      name="email"
      onChange={[Function]}
      placeholder="Enter Email"
      type="text"
      value=""
    />
    <input
      className="text"
      name="password"
      onChange={[Function]}
      placeholder="Enter Password"
      type="password"
      value=""
    />
    <button
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#9e0606",
          "borderRadius": "4px",
          "borderWidth": 0,
          "boxShadow": "0 3px 9px rgb(0 0 0 / 5%)",
          "color": "#fff",
          "cursor": "pointer",
          "display": "block",
          "fontSize": "1.4rem",
          "fontWeight": "bold,",
          "lineHeight": 1,
          "marginTop": "30px",
          "padding": "15px",
          "textAlign": "center",
          "transition": "all 180ms",
          "width": "100%",
        }
      }
      type="button"
    >
      Register
    </button>
    <div
      style={
        Object {
          "display": "flex",
          "justifyContent": "flex-end",
          "marginTop": "30px",
        }
      }
    >
      <a
        href="/login"
        onClick={[Function]}
        style={
          Object {
            "color": "#62b5e5",
            "textDecoration": "underline",
          }
        }
      >
        Login
      </a>
    </div>
  </div>
</div>
`;
